### DISCUSSION TITLE
Remap max pen pressure to be an eraser

### DISCUSSION THREAD

**Discussion by Aegeontis** [flxzt/rnote#1234] [Q&A] [answered]
I am using a Chromebook with a Penoval USI702 pen. This pen model has a tail eraser. In rnote, both the tip and tail are recognized as the same input. `sudo libinput debug-events` also sees both as the same input.
HOWEVER, the tail input always has the pressure 1.0, while the tip can vary between 0.0 and 1.0. It is rather difficult to get a 1.0 pressure with the tip, as I have to push it pretty hard into the screen.

So the solution would be to just remap `TABLET_TOOL_AXIS` and/or `TABLET_TOOL_TIP` with pressure 1.0 to be an eraser button.

Does anyone know how to do that? [Keyd](https://github.com/rvaiya/keyd) recognizes the pen, but doesn't seem to recognize the pressure levels.

**Comment by Doublonmousse** [comment]
Can you run rnote with `RUST_LOG=trace rnote` (or the equivalent command if it's from flatpak : flatpak --env=RUST_LOG=rnote=trace --env=RUST_BACKTRACE=1 run com.github.flxzt.rnote) and look at the output with the pen tip vs eraser tip (pressure is displayed in the trace)

That way you can verify this behavior is still the same as seen from rnote (to be sure the 1.0 isn't a default value for the absence of pressure data for libinput and things wouldn't work the same way in rnote).

**Reply by Aegeontis** [reply]
The pressure shown in rnote's logs is almost the same, except instead of being 1.00, it is now 0.999 for the eraser and still varies for the pen.

Technically the eraser also has pressure levels, but the eraser button has a very small travel distance and therefore the pressure curve almost instantly goes to 1.00 in libinput:
```
 event7   TABLET_TOOL_PROXIMITY   +28.467s		141.52*/150.11*	pressure: 0.00*	pen      (0xff, id 0) proximity-in 	axes:p	btn:S
 event7   TABLET_TOOL_TIP         +28.575s		142.57*/149.52*	pressure: 0.50* down
 event7   TABLET_TOOL_AXIS        +28.580s		142.57/149.52	pressure: 0.75*
 event7   TABLET_TOOL_AXIS        +28.588s		142.57/149.52	pressure: 0.87*
 event7   TABLET_TOOL_AXIS        +28.592s		142.57/149.52	pressure: 0.94*
 event7   TABLET_TOOL_AXIS        +28.600s		142.57/149.52	pressure: 0.97*
 event7   TABLET_TOOL_AXIS        +28.604s		142.57/149.52	pressure: 0.98*
 event7   TABLET_TOOL_AXIS        +28.611s		142.57/149.52	pressure: 0.99*
 event7   TABLET_TOOL_AXIS        +28.616s		142.57/149.52	pressure: 1.00*
 event7   TABLET_TOOL_AXIS        +28.624s		142.57/149.52	pressure: 1.00*
 event7   TABLET_TOOL_AXIS        +28.628s		142.57/149.52	pressure: 1.00*
 event7   TABLET_TOOL_AXIS        +28.636s		142.57/149.52	pressure: 1.00*
 event7   TABLET_TOOL_AXIS        +28.640s		142.57/149.52	pressure: 1.00*
 event7   TABLET_TOOL_AXIS        +28.648s		142.57/149.52	pressure: 1.00*
 event7   TABLET_TOOL_AXIS        +28.652s		142.57/149.52	pressure: 1.00*
 event7   TABLET_TOOL_AXIS        +28.660s		142.57/149.52	pressure: 1.00*
 event7   TABLET_TOOL_AXIS        +28.664s		142.57/149.52	pressure: 1.00*
 event7   TABLET_TOOL_AXIS        +28.672s		142.57/149.52	pressure: 1.00*
 event7   TABLET_TOOL_AXIS        +28.676s		142.57/149.52	pressure: 1.00*
 event7   TABLET_TOOL_AXIS        +28.684s		142.57/149.52	pressure: 1.00*
 event7   TABLET_TOOL_AXIS        +28.688s		142.57/149.52	pressure: 1.00*
 event7   TABLET_TOOL_AXIS        +28.696s		142.59*/149.48*	pressure: 1.00*
 event7   TABLET_TOOL_AXIS        +28.700s		142.67*/149.37*	pressure: 1.00*
 event7   TABLET_TOOL_AXIS        +28.708s		142.76*/149.24*	pressure: 0.75*
 event7   TABLET_TOOL_AXIS        +28.713s		142.86*/149.11*	pressure: 0.37*
 event7   TABLET_TOOL_AXIS        +28.732s		142.97*/149.05*	pressure: 1.00*
 event7   TABLET_TOOL_AXIS        +28.736s		143.06*/149.06*	pressure: 0.49*
 event7   TABLET_TOOL_TIP         +28.760s		143.25*/149.17*	pressure: 0.00* up
 event7   TABLET_TOOL_PROXIMITY   +28.762s		143.25/149.17	pressure: 0.00	pen      (0xff, id 0) proximity-out 
```

**Reply by Doublonmousse** [reply]
I'd like you to open an issue (maybe on the libinput side or appropriate repo that would have some involvement in the driver implementation for your pen) to (hopefully) getting it fixed upstream of rnote.

I wonder if we could have a "force touch" like thing that activates actions based on pen pressure (other than being a lone workaround for a single pen model).
I think a hold mode #728 would need to be there first though.

The issue currently is that things work well only if you have a (proximity in, pen down, pen moving, pen up, proximity out) sequence of the same pen type (tools have to be activated twice essentially and skipping a step won't work)

**Reply by Aegeontis** [reply]
I don't think libinput or the driver is really at fault here. According to this [reddit post](https://reddit.com/r/chromeos/comments/rw9rdy) the eraser part only works with specific models.

I have come up with an external workaround and will post a link here once I'm done with it.

**Comment by Aegeontis** [comment, marked as answer]
I created a small tool to fix this issue. To learn more head over to: https://github.com/Aegeontis/penoval-usi702-eraser-fix
